tools/draincluster/drain_test.go

```diff
@@ -0,0 +1,492 @@
+						TargetCluster: "test-cluster1",
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-crb2",
+						Labels: map[string]string{
+							placementv1beta1.CRPTrackingLabel: "test-crp1",
+						},
+					},
+					Spec: placementv1beta1.ResourceBindingSpec{
+						TargetCluster: "test-cluster2",
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-crb3",
+						Labels: map[string]string{
+							placementv1beta1.CRPTrackingLabel: "test-crp2",
+						},
+					},
+					Spec: placementv1beta1.ResourceBindingSpec{
+						TargetCluster: "test-cluster1",
+					},
+				},
+			},
+			wantErr: nil,
+			wantMap: map[string]bool{
+				"test-crp1": true,
+				"test-crp2": true,
+			},
+		},
+		{
+			name:          "no CRPs to evict",
+			targetCluster: "test-cluster1",
+			bindings: []placementv1beta1.ClusterResourceBinding{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-crb1",
+						Labels: map[string]string{
+							placementv1beta1.CRPTrackingLabel: "test-crp1",
+						},
+					},
+					Spec: placementv1beta1.ResourceBindingSpec{
+						TargetCluster: "test-cluster2",
+					},
+				},
+			},
+			wantErr: nil,
+			wantMap: map[string]bool{},
+		},
+		{
+			name:          "binding missing CRP label",
+			targetCluster: "test-cluster1",
+			bindings: []placementv1beta1.ClusterResourceBinding{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-crb1",
+					},
+					Spec: placementv1beta1.ResourceBindingSpec{
+						TargetCluster: "test-cluster1",
+					},
+				},
+			},
+			wantErr: fmt.Errorf("failed to get CRP name from binding test-crb1"),
+			wantMap: map[string]bool{},
+		},
+		{
+			name:          "skip CRB with deletionTimestamp",
+			targetCluster: "test-cluster1",
+			bindings: []placementv1beta1.ClusterResourceBinding{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-crb1",
+						Labels: map[string]string{
+							placementv1beta1.CRPTrackingLabel: "test-crp1",
+						},
+						DeletionTimestamp: &metav1.Time{Time: time.Now()},
+						Finalizers:        []string{"test-finalizer"},
+					},
+					Spec: placementv1beta1.ResourceBindingSpec{
+						TargetCluster: "test-cluster1",
+					},
+				},
+			},
+			wantErr: nil,
+			wantMap: map[string]bool{},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var objects []client.Object
+			scheme := serviceScheme(t)
+			for i := range tc.bindings {
+				objects = append(objects, &tc.bindings[i])
+			}
+			fakeClient := fake.NewClientBuilder().
+				WithScheme(scheme).
+				WithObjects(objects...).
+				Build()
+			h := helper{
+				hubClient:   fakeClient,
+				clusterName: tc.targetCluster,
+			}
+			gotMap, gotErr := h.fetchClusterResourcePlacementNamesToEvict(context.Background())
+			if tc.wantErr == nil {
+				if gotErr != nil {
+					t.Errorf("fetchClusterResourcePlacementNamesToEvict test %s failed, got error %v, want error %v", tc.name, gotErr, tc.wantErr)
```

Comment:
Hi Arvind! Style concerns aside, there are a few branches that really do not need to continue once failure are found (even though they can proceed).

Reply:
Something like 

```
if wantErred
    if gotErr != nil -> Errorf and return
    compareErrMessages -> Errorf if needed and return

if gotErr -> Errorf and return

cmpMap
```

would be clearer I assume.

Reply:
This would also apply to other cases I guess.